homeassistant/components/mqtt/fan.py

```diff
@@ -217,19 +362,91 @@ def state_received(msg):
+
+        if self._topic[CONF_PERCENTAGE_STATE_TOPIC] is not None:
+            topics[CONF_PERCENTAGE_STATE_TOPIC] = {
+                "topic": self._topic[CONF_PERCENTAGE_STATE_TOPIC],
+                "msg_callback": percentage_received,
+                "qos": self._config[CONF_QOS],
+            }
+            self._percentage = None
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def preset_mode_received(msg):
+            """Handle new received MQTT message for preset mode."""
+            preset_mode = self._templates[ATTR_PRESET_MODE](msg.payload)
+            if preset_mode in self.preset_modes:
+                self._preset_mode = preset_mode
+                if not self._implemented_percentage and (
+                    preset_mode in self.speed_list
+                ):
+                    self._percentage = ordered_list_item_to_percentage(
+                        self.speed_list, preset_mode
+                    )
+            else:
+                raise NotValidPresetModeError(
+                    f"{msg.payload} is not a valid preset mode"
+                )
+
+            self.async_write_ha_state()
+
+        if self._topic[CONF_PRESET_MODE_STATE_TOPIC] is not None:
+            topics[CONF_PRESET_MODE_STATE_TOPIC] = {
+                "topic": self._topic[CONF_PRESET_MODE_STATE_TOPIC],
+                "msg_callback": preset_mode_received,
+                "qos": self._config[CONF_QOS],
+            }
+            self._preset_mode = None
+
+        # The use of legacy speeds is deprecated in the schema, support will be removed after a quarter (2021.7)
         @callback
         @log_messages(self.hass, self.entity_id)
         def speed_received(msg):
             """Handle new received MQTT message for the speed."""
-            payload = self._templates[ATTR_SPEED](msg.payload)
-            if payload == self._payload["SPEED_LOW"]:
-                self._speed = SPEED_LOW
-            elif payload == self._payload["SPEED_MEDIUM"]:
-                self._speed = SPEED_MEDIUM
-            elif payload == self._payload["SPEED_HIGH"]:
-                self._speed = SPEED_HIGH
-            elif payload == self._payload["SPEED_OFF"]:
-                self._speed = SPEED_OFF
+            speed_payload = self._templates[ATTR_SPEED](msg.payload)
+            if speed_payload == self._payload["SPEED_LOW"]:
+                speed = SPEED_LOW
+            elif speed_payload == self._payload["SPEED_MEDIUM"]:
+                speed = SPEED_MEDIUM
+            elif speed_payload == self._payload["SPEED_HIGH"]:
+                speed = SPEED_HIGH
+            elif speed_payload == self._payload["SPEED_OFF"]:
+                speed = SPEED_OFF
+            else:
+                speed = None
+
+            if speed and speed in self._legacy_speeds_list:
+                self._speed = speed
+            else:
+                raise NotValidSpeedError(f"'{msg.payload}' is not a valid speed.")
```

Comment:
This doesn't look very nice in the log:
```
2021-03-25 08:26:01 ERROR (MainThread) [homeassistant.util.logging] Exception in speed_received when handling msg on 'stat/bvent/SPEED': '100'
Traceback (most recent call last):
  File "/home/erik/development/home-assistant_fork/homeassistant/components/mqtt/debug_info.py", line 32, in wrapper
    msg_callback(msg)
  File "/home/erik/development/home-assistant_fork/homeassistant/components/mqtt/fan.py", line 440, in speed_received
    raise NotValidSpeedError(f"'{msg.payload}' is not a valid speed.")
homeassistant.components.fan.NotValidSpeedError: '100' is not a valid speed.
```

Log a warning instead, then return:
```py
# At top level:
import logging
...
_LOGGER = logging.getLogger(__name__)

# In speed_received:
_LOGGER.warning(
    "'%s' received on topic %s is not a valid speed", msg.payload, msg.topic
)
return
```

Resulting log:
```
2021-03-25 08:37:20 WARNING (MainThread) [homeassistant.components.mqtt.fan] '100' received on topic stat/bvent/SPEED is not a valid speed
```

Same comment for other exceptions raised when handling MQTT messages.

In the tests you can do:
```
async_fire_mqtt_message(hass, "speed-state-topic", "speed_mEdium")
assert "not a valid speed" in caplog.text
caplog.clear()
```

Reply:
I have replaced the code where exceptions where raised with logging a warning as you suggested. Further I have added some additional tests on alternate speed_range configurations.